### PR TITLE
Use coarse grained memory reporting to reduce congestion

### PR DESCRIPTION
**Pull request by ShubhamChaurasia**
## Description
This reduces congestion in the memory tracking system by coarsening the granularity of memory tracking. In this paradigm, the most small incremental increases at the byte granularity will not actually result in a different coarse granularity value, so no reporting into memory tracking is necessary.
This has already been addressed previously in https://github.com/trinodb/trino/pull/22668, however it under reports memory to the tracking system which can be a problem in high concurrency environment.

In this approach, we prefer to over-report than under report within the small rounding margin. Also, this eliminates the need of an explicit `sync()` API.

Thanks to @pettyjamesm for the idea and the guidance for this.

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Review by pettyjamesm — COMMENTED** [review]
LGTM, except for some minor comments. cc: @findepi - this avoids under-reporting the amount of memory used which can be dangerous at high concurrency.

File: lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java [hunk 1]
```diff
@@ -14,32 +14,40 @@
...
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
-public class ThresholdLocalMemoryContext
+/**
+ *  This class prevents contention in the memory tracking system by coarsening the granularity of memory tracking.
+ *  In this paradigm, the most small incremental increases at the byte granularity will not actually result
+ *  in a different coarse granularity value, so no reporting into memory tracking is necessary
+ */
+public class CoarseGrainLocalMemoryContext
```
Comment (sopel39) [inline lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java:29]: All `LocalMemoryContext` implementations are thread safe. I think this one should be too cc @losipiuk

**Comment by ShubhamChaurasia** [comment]
thanks for the review @pettyjamesm  @sopel39 -- addressed the comments.

File: lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+    public CoarseGrainLocalMemoryContext(LocalMemoryContext delegate)
+    {
+        this(delegate, DEFAULT_GRANULARITY);
+    }
+
+    public CoarseGrainLocalMemoryContext(LocalMemoryContext delegate, long granularity)
+    {
+        this.delegate = requireNonNull(delegate, "delegate is null");
+        checkArgument(granularity > 0, "syncThreshold must be greater than 0");
```
Comment (pettyjamesm) [inline lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java:47]: `"granularity must be greater than 0"`

**Review by pettyjamesm — APPROVED** [review]
LGTM, @losipiuk PTAL

**Review by losipiuk — APPROVED** [review]
Thanks. Looks good :)

File: lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+ */
+public class CoarseGrainLocalMemoryContext
+        implements LocalMemoryContext
+{
+    public static final long DEFAULT_GRANULARITY = 65536;
+
+    private final LocalMemoryContext delegate;
+    private final long granularity;
+    private final long mask;
+    private long currentBytes;
```
Comment (sopel39) [inline lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java:37]:
```
    @GuardedBy("this")
```
is missing

File: lib/trino-memory-context/src/main/java/io/trino/memory/context/ThresholdLocalMemoryContext.java [hunk 4]
```diff
@@ -1,93 +0,0 @@
-/*
```
Comment (sopel39) [inline lib/trino-memory-context/src/main/java/io/trino/memory/context/ThresholdLocalMemoryContext.java:1]: nit: this should be two commits:
1. rename
2. changes on top
this way history is preserved

File: lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java (hunk 3 above)
Comment (pettyjamesm) [reply lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java:37]: Added in https://github.com/trinodb/trino/pull/22881

File: lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+
+/**
+ *  This class prevents contention in the memory tracking system by coarsening the granularity of memory tracking.
```
Comment (breglend) [inline lib/trino-memory-context/src/main/java/io/trino/memory/context/CoarseGrainLocalMemoryContext.java:25]: nit: Possible typo (description says "congestion," here we say "contention").

File: lib/trino-memory-context/src/test/java/io/trino/memory/context/TestMemoryContexts.java [hunk 6]
```diff
@@ -158,6 +158,66 @@ public void testClosedLocalMemoryContext()
...
+        assertCoarseGrainContextValues(coarseGrainContext, delegate, 2 * granularity, 2 * granularity);
+
+        // gets set to the next coarse unit
+        // k*granularity + x in [1, granularity] leads to setting the context to (k+1) * granularity
+        assertCoarseGrainContextValues(coarseGrainContext, delegate, 0, 0);
+        for (int i = 1; i <= granularity; i++) {
+            assertCoarseGrainContextValues(coarseGrainContext, delegate, i, granularity);
+        }
+
+        // threshold not a power of 2
```
Comment (breglend) [inline lib/trino-memory-context/src/test/java/io/trino/memory/context/TestMemoryContexts.java:189]: nit: "granularity"

**Comment by colebow** [comment]
Does this need release notes? cc @pettyjamesm

**Comment by pettyjamesm** [comment]
Sorry about that @colebow, I should have verified the release notes section. No notes should be necessary on this one.

**Pull request merged**